DataFrame-based image reader and transformer

hhbyyh opened the pull request:
## What changes were proposed in this pull request?

Provides DataFrame-based image reader and feature transformer. 
Both Scala and Python interface are provided.

More unit tests will be added.

## How was this patch tested?

new unit tests.

hhbyyh commented:
Please help review @songhappy

jason-dai commented on spark/dl/src/main/scala/com/intel/analytics/bigdl/dlframes/DLImageReader.scala:
```diff
@@ -0,0 +1,143 @@
+      StructField("data", new ArrayType(FloatType, false), false) :: Nil)
+
+  private[dlframes] def imf2Row(imf: ImageFeature): Row = {
+    val (mode, data) = if (imf.contains(ImageFeature.imageTensor)) {
```
need to handle ImageFeature.floats as well

jason-dai commented on spark/dl/src/main/scala/com/intel/analytics/bigdl/dlframes/DLImageTransformer.scala:
```diff
@@ -0,0 +1,93 @@
+
+    val transformUDF = SqlAdapter.getUDF(transFunc, DLImageSchema.floatSchema)
+
+    dataFrame.withColumn($(outputCol), transformUDF(dataFrame($(inputCol))))
```
1. Use the transform in Transformer or FeatuerTransformer instead, as it will handle broadcast, clone, etc.; someting like:
   ```
   r1 = dataFrame.rdd().map(DLImageSchema.row2IMF(_)))
   r2 = localTransformer(r1).map(DLImageSchema.imf2Row(_))
   ```

2. Do we want to replicate the image contents in both inputCol and outputCol here?

hhbyyh replied:
I'm thinking we should support the case the there're multiple columns in the input DataFrame. To use the RDD interface, we need to extract, transform then zip back with the other columns, which can bring extra performance cost and instability during runtime.

For 2, user may set inputCol and outputCol to the same name. I'll add a unit test.

jason-dai replied:
The current usage of transformer in the code is incorrect; you also need to clone the transformer after broadcast; see https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala#L67

hhbyyh replied:
Thanks for the reminder. I'll update and run some tests.

jason-dai commented:
@wzhongyuan please take a look

wzhongyuan commented on spark/dl/src/main/scala/com/intel/analytics/bigdl/dlframes/DLImageTransformer.scala:
```diff
@@ -0,0 +1,93 @@
+  override def transformSchema(schema: StructType): StructType = {
+    val inputType = schema($(inputCol)).dataType
+    validateInputType(inputType)
+    if (schema.fieldNames.contains($(outputCol))) {
```
So each transformer in the chain will have its own output column ?  that means the data will be duplicated.

hhbyyh replied:
Only one output column for the chained transformers. Please check the unit test for example

wzhongyuan commented on spark/dl/src/main/scala/com/intel/analytics/bigdl/dlframes/DLImageReader.scala:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Definition for image data in a DataFrame
+ */
+object DLImageSchema {
```
make it private to  package `dlframes` ?

hhbyyh replied:
Though it may not be the major use case, but it's possible that user want to manually compose a image column.
The byteSchema is public in Spark Image schema so we probably will keep it stable.

wzhongyuan replied:
Makes sense

hhbyyh commented:
Any comments?

wzhongyuan commented on pyspark/test/bigdl/dlframes/test_dl_image_reader.py:
```diff
@@ -0,0 +1,58 @@
+    def setup_method(self, method):
+        """
+        setup any state tied to the execution of the given method in a
+        class.  setup_method is invoked for every test method of a class.
```
remove unnecessary space ?

wzhongyuan commented on pyspark/test/bigdl/dlframes/test_dl_image_reader.py:
```diff
@@ -0,0 +1,58 @@
+
+    def teardown_method(self, method):
+        """
+         teardown any state that was previously setup with a setup_method
```
indent ?

hhbyyh commented:
Thanks for the comments. Fixed the format issue and rebased.

PR validation passed: http://172.168.2.101:8080/view/PR-V(new)-Pipeline/job/BigDL-PR-Validation-v3/341/

hhbyyh commented:
Updated the code to use mapPartitions and clone.

jason-dai commented:
LGTM

hhbyyh commented:
http://172.168.2.101:8080/view/PR-V(new)-Pipeline/job/BigDL-PR-Validation-v3/365/ passed. Merging to master
